### PR TITLE
Increase HIP compatibility 

### PR DESCRIPTION
**Pull request by psychocoderHPC**
Increase HIP compatibility to PMacc and PIConGPU.

- HIP support for RNG XorMin
- update pre-compiler switches to support HIP

This PR is not fully introduce HIP support. There will be more updates soon.

### CODE REVIEW

File: include/picongpu/fields/currentDeposition/Strategy.def [hunk 1]
```diff
@@ -147,6 +147,18 @@ namespace traits
     };
 #endif
 
+#if( ALPAKA_ACC_GPU_HIP_ENABLED == 1 )
+        template<
```
Comment (sbastrakov) [inline include/picongpu/fields/currentDeposition/Strategy.def:151]: This line is misaligned

Comment (sbastrakov) [reply include/picongpu/fields/currentDeposition/Strategy.def:151]: Not sure, perhaps we could add a short description of why HIP a different default strategy?

File: include/picongpu/particles/ParticlesFunctors.hpp [hunk 2]
```diff
@@ -126,7 +126,7 @@ struct LogMemoryStatisticsForSpecies
         const std::shared_ptr<T_DeviceHeap>& deviceHeap
     ) const
     {
-#if( PMACC_CUDA_ENABLED == 1 )
+#if( BOOST_LANG_CUDA || BOOST_COMP_HIP)
```
Comment (sbastrakov) [inline include/picongpu/particles/ParticlesFunctors.hpp:129]: Maybe we could introduce smth like  `PMACC_CUDA_OR_HIP_ENABLED` to express that `BOOST_LANG_CUDA || BOOST_COMP_HIP` ? This expression is repeating in several placed

File: include/pmacc/Environment.hpp [hunk 3]
```diff
@@ -490,9 +490,11 @@ namespace detail
             const int tryDeviceId = (deviceOffset + deviceNumber) % num_gpus;
 
             log<ggLog::CUDA_RT>("Trying to allocate device %1%.") % tryDeviceId;
+
 #if (PMACC_CUDA_ENABLED == 1)
             cudaDeviceProp devProp;
-            CUDA_CHECK((cuplaError_t)cudaGetDeviceProperties(&devProp, tryDeviceId));
+
+            CUDA_CHECK((cuplaError_t)ALPAKA_API_PREFIX(GetDeviceProperties)(&devProp, tryDeviceId));
```
Comment (sbastrakov) [inline include/pmacc/Environment.hpp:497]: Should this code stay only for CUDA? (I do not know, just wondering if that is now supposed to be CUDA and HIP)

File: include/pmacc/Environment.hpp [hunk 4]
```diff
@@ -527,17 +529,16 @@ namespace detail
             if (rc == cuplaSuccess)
             {
 #if (PMACC_CUDA_ENABLED == 1)
-                cudaDeviceProp dprop;
-                CUDA_CHECK((cuplaError_t)cudaGetDeviceProperties(&dprop, tryDeviceId));
-                log<ggLog::CUDA_RT> ("Set device to %1%: %2%") % tryDeviceId % dprop.name;
+                CUDA_CHECK((cuplaError_t)ALPAKA_API_PREFIX(GetDeviceProperties)(&devProp, tryDeviceId));
+                log<ggLog::CUDA_RT> ("Set device to %1%: %2%") % tryDeviceId % devProp.name;
```
Comment (sbastrakov) [inline include/pmacc/Environment.hpp:533]: Same as previous comment, should this stay CUDA-only?

File: include/picongpu/particles/ParticlesFunctors.hpp (hunk 2 above)
Comment (sbastrakov) [reply include/picongpu/particles/ParticlesFunctors.hpp:129]: Discussed offline: this is not needed.

File: include/pmacc/algorithms/math/doubleMath/floatingPoint.tpp [hunk 5]
```diff
@@ -69,7 +69,7 @@ struct Float2int_rn<double>
 
     HDINLINE result operator( )(double value)
     {
-#if __CUDA_ARCH__
+#if( defined(__CUDA_ARCH__) || __HIP_DEVICE_COMPILE__ == 1 ) //we are on gpu
```
Comment (sbastrakov) [inline include/pmacc/algorithms/math/doubleMath/floatingPoint.tpp:72]: This needs to be updated

File: include/pmacc/math/ConstVector.hpp [hunk 6]
```diff
@@ -26,13 +26,13 @@
 #include "pmacc/types.hpp"
 
 /* select namespace depending on __CUDA_ARCH__ compiler flag*/
-#ifdef __CUDA_ARCH__ //we are on gpu
+#if( defined(__CUDA_ARCH__) || __HIP_DEVICE_COMPILE__ == 1 )  //we are on gpu
```
Comment (sbastrakov) [inline include/pmacc/math/ConstVector.hpp:29]: Here too

File: include/pmacc/Environment.hpp (hunk 4 above)
Comment (psychocoderHPC) [reply include/pmacc/Environment.hpp]: fixed by activating this code for HIP too

File: include/picongpu/fields/currentDeposition/Strategy.def (hunk 1 above)
Comment (psychocoderHPC) [reply include/picongpu/fields/currentDeposition/Strategy.def]: Add a short discription

File: include/pmacc/random/methods/XorMin.hpp [hunk 7]
```diff
@@ -38,15 +40,17 @@ namespace random
...
-    using XorMin = AlpakaRand< T_Acc >;
-#else
+#if( ALPAKA_ACC_GPU_CUDA_ENABLED || ALPAKA_ACC_GPU_HIP_ENABLED )
     //! Uses the CUDA XORWOW RNG but does not store state members required for normal distribution
     template< typename T_Acc = cupla::Acc>
     class XorMin
     {
+#if (BOOST_LANG_HIP)
+        using NativeStateType = hiprandStateXORWOW_t;
+#elif (BOOST_LANG_CUDA)
```
Comment (sbastrakov) [inline include/pmacc/random/methods/XorMin.hpp:50]: These brackets do not follow the style of the surrounding code.

**Review by sbastrakov — DISMISSED** [review]
Nice, only a small formatting suggestion. Also CI failed due to some internal reasons I guess

**Comment by psychocoderHPC** [comment]
I pushed the change, hope clang can now be downloaded by the ci

**Review by sbastrakov — APPROVED** [review]

**Pull request merged**